keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+
+```yaml
+# Policy configurations
+apiVersion: rules.example.com/v1
+kind: ReplicaLimit
+metadata:
+  name: "replica-limit-prod.example.com"
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: In,
+      values: ["prod"]
+  maxReplicas: 100
+```
+
+This design separates admission policy _definition_ from _configuration_. This
+has a couple advantages:
+
+- Access to, and delegation of, policy configuration is more manageable. In
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Constraints
+
+Each `ValidatingAdmissionPolicy` resource may optionally set `spec.constraints`
+to constrain the resources it validates. 
+
+- `spec.constraints` limits which resources may be matched configurations of the
+  policy. This allows the CEL expressions to make safe assumptions. E.g. a CEL
+  expression that is constrained to CREATES and UPDATES of resources is
+  guaranteed the root `object` variable is never null, but a CEL expression that
+  might need to evaluate a DELETE must handle the root `object` variable being
+  null.
+- `spec.constraints` is used for CEL expression type checking, see the below
+  "Type safety" section for more details.
+
+
+Alternatives considered:
+
+- Name the constraint field `spec.constraints` instead of `spec.match`. This
+  field name would imply that the field works the same as a policy configuration
+  object's `match` field, but we intend for different semantics here. We're
+  setting constraints on what can be matched, not defining match criteria
+  directly.
+- Use a GVK instead of match rules. This works well for type checking, but doesn't
+  help establish a the criteria for what resources a policy is allowed to apply to.
+
+##### Configuration
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
+"configuration CRDs" used to configure the admission policy.
+
+Policy configuration CRDs are the interface between `ValidatingAdmissionPolicy`
+authors and the administrators that configure the policies for clusters. The
+configuration CRD allows policy definition authors to define how exactly a
+`ValidatingAdmissionPolicy` may be configured using a OpenAPIv3 structural
+schema.
+
+Policy configuration CRDs have a few restrictions:
+
+- Match criteria fields must be defined and must confirm to the expected schema.
+  See "Match Criteria" for more details.
+- For phase 1, configuration CRDs must be cluster scoped. See phase 2 (below) for
+  our plan on namespace scoped configuration CRDs.
+
+If any of the above configuration CRD restrictions are violated, the errors will
+be reported in the status of the `ValidatingAdmissionPolicy`. If the match
+criteria is malformed, this unfortunately may cause the policy to fail open--
```

Comment:
is fail open versus fail closed an option in the ValidatingAdmissionPolicy?

Reply:
Yes, I was thinking of putting the `failureMode` flag (defaulting to `Fail`) on each validation expression, that way they could be independently controlled without having to split out the expressions into two policy resources.

It occurred to me that there might be reasons to have `failureMode` as a field in the configuration resources, but I haven't thought it through carefully.